### ISSUE TITLE
Is styling div:.phpdebugbar * appropriate

### ISSUE THREAD

**Issue by bobd91** [maximebf/php-debugbar#205]
The latest debugbar.css contains:

div.phpdebugbar * {
  margin: 0;
  padding: 0;
  border: 0;
  font-weight: normal;
  text-decoration: none;
  clear: initial;
  width: auto;
  -moz-box-sizing: content-box;
       box-sizing: content-box;
}
(In the previous version I was using it just affected box-sizing which didn't affect me too much.)

Most of the selectors in my CSS are single class selectors and they are now just being ignored.  I could go through and make them all more specific but I shouldn't have to.  And how can I unapply that clear setting (what is clear: initial anyway?) 

Given that the debugbar is meant to contain components from other authors is this appropriate?

Thanks,
Bob

**Comment by barryvdh** [comment]
What styling does this effect? You created your own widgets?

This is mostly meant to reset inherited styling from the original document.

**Comment by bobd91** [comment]
I have my own widget that that has a tab on the bar and content in the phpdebugbar-panel.

If I have a component with class 'bobs-widget' and style it with

.bobs-widget { margin: 1px; padding: 2px; border: 3px; width: 4px; ... }

my styles are ignored because the 'div.phpdebugbar *' selector is more specific.

I just did a quick test and changing 'div.phpdebugbar *' to '.phpdebugbar *' would fix things for me, would that still achieve what you require?

**Comment by barryvdh** [comment]
I'm not sure, could have been necesarry to increate the specificity. Can't you prefix your styles with div.phpdebugbar?

**Comment by bobd91** [comment]
Yes, I guess so.  Other downstream widget builds might be similarly affected when they get your Feb changes.

**Comment by bobd91** [comment]
Hi Barry,

re. resetting inherited styles; shouldn't that be as unspecific as possible?

Shouldn't it say 
'if you haven't set this style then please inherit this setting rather than one from the page'
rather than 
'Use this setting even if you have chosen a different one'.

Regards,
Bob

**Comment by barryvdh** [comment]
So how do you propose to do that?

**Comment by bobd91** [comment]
Wouldn't  '.phpdebugbar *' do the job?

Any nested elements would have those styles applied if they didn't set them.  Any styling on nested elements would only need a single class selector to override as long as they are later in the load order.

**Comment by geekwright** [comment]
I'll just chime in here.  I really appreciate the current specificity.

I remember earlier days when there was lots of bleed over from the application styles, making for unpredictable behaviors. Adding the full reset starts everything from a nice consistent point and has eliminated the interference. Loosening that selector would increase the chance that something outside could outweigh it.

I'd would much rather have to add a little bit to custom selectors than to have to deal with less predictability.

**Comment by bobd91** [comment]
You mention 'earlier days' but the breaking changes only went in in February.  Are you aware of improvements since then?

(The selector may have been there for a while but until February it was only used to enforce box-sizing)

**Comment by geekwright** [comment]
I didn't analyze the impact of individual rules, but 1.10.4 is the most visually consistent yet.

**Comment by bobd91** [comment]
I changed my styles to work with this, I have even added borders back to my drop down lists ;)

But the show stopper has been the use of a third party component (Ace editor) which didn't realise it has to be so specific with its styles.  There is no way I am going to fork that project to make its styles compatible with phpdebugbar.

For the moment I am going to have to work off a phpdebugbar fork.  I hope that enough people find this choice of css inappropriate and it gets changed so I can come back to the main project.

**Comment by barryvdh** [comment]
what is the link to the repo of your collector?

**Comment by barryvdh** [comment]
And what css line is exactly the problem? Like box sizing, perhaps I can set the style just for the parent div and set the rest to inherit, would that help?

**Comment by bobd91** [comment]
https://github.com/bobd91/magedebugbar
it needs a Magento eCommerce installation to work, you can see it running at
http://demo.magedebugbar.org

the css that showed obvious problems and made me give up was the forced 'width: auto;' as some widths in the Ace editor were clearly being ignored.

I had issues with border: 0; as it was removing borders from elements that normally have them (selects for example) but I could get round that by reapplying them though I didn't know what the default border colour was.

There maybe other related issues but now I have removed the div from div.phpdebugar * everything seems ok.

**Comment by barryvdh** [comment]
Okay cool, I had actually started playing with a magento version a few weeks back also, so will take a look.

You could af2f your own css to overwrite some css, but you can also remove the original css file I think.

**Comment by bobd91** [comment]
Re: Magento - my stuff only handles layout at the moment, though I do hope to add more.

Re: af2f etc. - sorry but I don't understand.

**Comment by barryvdh** [comment]
I'll try to take a look soon if there is a better way to prevent style bleeding from the website to the debugbar.
